### ISSUE TITLE
twisted.conch.telnet dies in conjunction with PuTTY

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1967]
|<img alt="sam's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| sam reported|
|-|-|
|Trac ID|trac#1967|
|Type|defect|
|Created|2006-07-28 17:32:53Z|

twisted.conch.telnet crashes when PuTTY is connected to it on Windows XP SP2, Python 2.5b1, twisted 2.4 (but twisted.conch.telnet at rev 17656). Minimal crash example:

```python
from twisted.conch.telnet import Telnet
from twisted.internet.protocol import Factory
from twisted.internet import reactor

f = Factory()
f.protocol = Telnet

reactor.listenTCP(2323, f)
reactor.run()
```

and then from a shell with plink, PuTTY's commandline tool, visible:

```$ plink -telnet -P 2323 localhost```

The problem appears to originate in lines 664-674 of twisted.conch.telnet. I can't figure out why those methods are calling self.protocol.*, because there's no reference to self.protocol in the rest of the Telnet class. There is, however, an identical definition a few lines down, in the TelnetTransport class, where using self.protocol makes sense.

Changing the definition of those four functions in Telnet to "pass" would fix this problem, but I don't know if this is an acceptable solution. They are, presently, unsuitable for conventional use, so I doubt that this would break code, but it may break the defined behaviour of these methods.

<details><summary>Searchable metadata</summary>

```
trac-id__1967 1967
type__defect defect
reporter__sam sam
priority__highest highest
milestone__Conch_0_9 Conch-0.9
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1154107973000000 1154107973000000
changetime__1190290619000000 1190290619000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```pass``` is basically right, I think, although ```return False``` would be more clear.  I think those implementations are a hold-over from a prior version which used fewer classes.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

Is there a test that should go with this?  Or is this simply functionality that shouldn't exist at all?  (If so, the path where it gets invoked should be tested to make sure it doesn't blow up...)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun removed owner|
|-|-|

Ready for review in telnet-option-negotiation-1967

**Comment by twisted-trac** [comment]
|<img alt="tjs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|tjs commented|
|-|-|

makes sense to me, tests pass, works in putty, I'd possibly consider passing the docstring about when to implement the disable methods to the error being raised but other than that its all good. btw you suck, I've now done something constructive for twisted :(

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [17829]) Merge telnet-option-negotiation-1967

Author: exarkun
Reviewer: tjs
Fixes [#1967](https://github.com/twisted/twisted/issues/1967)

This replaces the implementations of Telnet.{enable,disable}{Local,Remote}, which were
simply broken, with implementations which reject all option negotiations.  It also adds
test coverage for the behavior of these base implementations.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [19153]) Revert r17829

Refs [#1967](https://github.com/twisted/twisted/issues/1967)
Refs [#2099](https://github.com/twisted/twisted/issues/2099)

This changeset introduced a regression in untested code which manifests
when accessing a conch telnet server using the NetKit telnet client.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun removed owner|
|-|-|

See final comment on [#2099](https://github.com/twisted/twisted/issues/2099).

Code for this ticket is now in teloptneg-1967

It's very similar to the previous effort (although I implemented it tdd from scratch before I looked at the diff for the previous branch, which I hadn't looked at since the merge was reverted).  It does omit the changes to TelnetProtocol, though, which were not necessary to make the tests for the problematic behavior, enableLocal et al raising AttributeErrors, pass.  One could argue that they are incorrect and should be fixed, but one should do so with a failing unit test. ;)

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set owner to @exarkun|
|-|-|

 * Both telnet.py and test_telnet.py should have their copyright updated
 * telnet.py should get a test-case-name declaration
 * unused import internet in telnet.py

I've looked at the problem with [#2099](https://github.com/twisted/twisted/issues/2099): the changes to ```TelnetProtocol``` break the netkit client, so it seems wise not to put it again :). There are probably some stuff missing in ```TelnetTransport```, but that doesn't matter for now.

I've tested the behavior with Putty, and it looked good, so once the few changes above made, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

Done in r21296, r21297, and r21298.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [21299]) Merge teloptneg-1967

Author: exarkun
Reviewer: therve
Fixes [#1967](https://github.com/twisted/twisted/issues/1967)
Refs [#2099](https://github.com/twisted/twisted/issues/2099)

Change the base telnet protocol implementation, twisted.conch.telnet.Telnet,
so that, by default, it cleanly rejects all attempts to enable options, rather
than raising AttributeErrors which cause the connection to be lost.  Also
change it to raise NotImplementedError, rather than an AttributeError, to
indicate a programming error if one of its disable callbacks is somehow
invoked.

This allows simple Telnet subclasses which do not customize option negotiation
to work when used with clients which attempt option negotiation.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
